### PR TITLE
 Allow only the `error?` as return type when `http:Caller` is present as resource function arg

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes [#1519](https://github.com/ballerina-platform/ballerina-standard-library/issues/1519)

## Examples
Check following code-snippet for valid resource method declarations with `http:Caller` as a parameter.
```ballerina
service / on new http:Listener(9999) {
    // invalid resource method declaration
    resource function get invalid (http:Caller caller, string action) returns http:BadRequest? {
        // ....
    }
   
    // valid resource method declaration
    resource function get valid (string action) returns http:Accepted|http:BadRequest {
        // ....
    }
    
    // valid resource method declaration
    resource function get validWithCaller (http:Caller caller, string action) returns error? {
        // ...
    }
}
```

## Checklist
- [X] Linked to an issue
- [X] Updated the changelog
- [X] Added tests

### CODE REVIEW

File: http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import ballerina/http;
```
Comment (shafreenAnfar) [inline http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal:1]: License header

File: http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
```
Comment (shafreenAnfar) [inline http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal:3]: Additional new lines

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/CompilerPluginUtils.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package io.ballerina.stdlib.http.compiler;
```
Comment (shafreenAnfar) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/CompilerPluginUtils.java:1]: ditto

**Review by shafreenAnfar — APPROVED** [review]

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java [hunk 4]
```diff
@@ -52,7 +52,8 @@
             "to the 'http:CallerInfo' annotation", ERROR),
     HTTP_115("HTTP_115", "invalid multiple 'http:Caller' parameter: '%s'", ERROR),
     HTTP_116("HTTP_116", "invalid multiple 'http:Request' parameter: '%s'", ERROR),
-    HTTP_117("HTTP_117", "invalid multiple 'http:Headers' parameter: '%s'", ERROR);
+    HTTP_117("HTTP_117", "invalid multiple 'http:Headers' parameter: '%s'", ERROR),
+    HTTP_118("HTTP_118", "could not use 'http:Caller' and return '%s' from a resource", ERROR);
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java:56]: lets change the error to be aligned with other errors
```
invalid resource method return type: can not use 'http:Caller' and return '%s' from a resource : expected 'error?'
```

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 5]
```diff
@@ -536,6 +539,37 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
         return expectedType.equals(typeName.get());
     }
 
+    private static void validateHttpCallerUsage(SyntaxNodeAnalysisContext ctx, FunctionDefinitionNode member) {
+        Optional<Symbol> resourceMethodSymbolOptional = ctx.semanticModel().symbol(member);
+        if (resourceMethodSymbolOptional.isPresent()) {
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:544]: Let's do an early return here. So code column width can be reduced.
```
        if (resourceMethodSymbolOptional.isEmpty()) {
            return;
        }
```

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 6]
```diff
@@ -536,6 +539,37 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
...
+            ResourceMethodSymbol resourceMethod = (ResourceMethodSymbol) resourceMethodSymbolOptional.get();
+            FunctionTypeSymbol typeDescriptor = resourceMethod.typeDescriptor();
+            Optional<List<ParameterSymbol>> parameterOptional = typeDescriptor.params();
+            boolean isCallerPresent = parameterOptional.isPresent() &&
+                    parameterOptional.get().stream().anyMatch(HttpResourceValidator::isHttpCaller);
+            if (isCallerPresent) {
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:550]: Early return here as well

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 7]
```diff
@@ -536,6 +539,37 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
...
+                Optional<TypeSymbol> returnTypeDescriptorOpt = typeDescriptor.returnTypeDescriptor();
+                if (returnTypeDescriptorOpt.isPresent()) {
+                    TypeSymbol typeSymbol = returnTypeDescriptorOpt.get();
+                    String returnTypeDescription = constructReturnTypeDescription(typeSymbol);
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:554]: Why do we do string matching here? Ideally this should be a type check.

File: http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            return result;
+        } else {
+            http:BadRequest result = {
+                body: "Provided `action` parameter is invalid"
+            };
+            return result;
+        }
+    }
+
+    resource function get validWithCaller (http:Caller caller, string action) returns error? {
```
Comment (chamil321) [inline http-compiler-plugin-tests/src/test/resources/ballerina_sources/sample_package_13/service.bal:30]: let's add a test case for HTTP module errors as well

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 9]
```diff
@@ -538,6 +536,63 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
...
+    }
+
+    private static boolean isHttpCaller(ParameterSymbol param) {
+        TypeDescKind typeDescKind = param.typeDescriptor().typeKind();
+        if (TypeDescKind.TYPE_REFERENCE.equals(typeDescKind)) {
+            TypeSymbol typeDescriptor = ((TypeReferenceTypeSymbol) param.typeDescriptor()).typeDescriptor();
+            return isHttpModuleType(CALLER_OBJ_NAME, typeDescriptor);
+        } else {
+            return false;
+        }
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:572]:
```suggestion
        return false;
```

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 10]
```diff
@@ -538,6 +536,63 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
...
+    }
+
+    private static String getReturnTypeDescription(ReturnTypeDescriptorNode returnTypeDescriptorNode) {
+        Node returnTypeDescriptor = returnTypeDescriptorNode.type();
+        return returnTypeDescriptor.toString().split(" ")[0];
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:577]:
```suggestion
        return returnTypeDescriptorNode.type().toString().split(" ")[0];
```

File: http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 11]
```diff
@@ -538,6 +536,63 @@ private static boolean isHttpModuleType(String expectedType, TypeSymbol typeDesc
...
+        if (TypeDescKind.TYPE_REFERENCE.equals(typeDescKind)) {
+            TypeSymbol typeDescriptor = ((TypeReferenceTypeSymbol) param.typeDescriptor()).typeDescriptor();
+            return isHttpModuleType(CALLER_OBJ_NAME, typeDescriptor);
+        } else {
+            return false;
+        }
+    }
+
+    private static String getReturnTypeDescription(ReturnTypeDescriptorNode returnTypeDescriptorNode) {
+        Node returnTypeDescriptor = returnTypeDescriptorNode.type();
```
Comment (chamil321) [inline http-compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:576]:
```suggestion
```

**Review by chamil321 — APPROVED** [review]

**Review by chamil321 — APPROVED** [review]

**Pull request merged**